### PR TITLE
Pause playback on (bluetooth) headphones disconnect.

### PR DESCRIPTION
**Pull request by rotvel**
Fix https://github.com/dkanada/gelli/issues/61 by moving the BECOMING_NOISY logic to MultiPlayer.

### CODE REVIEW

**Comment by rotvel** [comment]
...while not breaking gapless playback.

**Comment by dkanada** [comment]
Is there any way to do this without moving it to the MultiPlayer class? It's a terrible name, because it's really just an implementation for the generic playback interface. If a SessionPlayer or ChromecastPlayer are ever added your bluetooth fix would only work for the ExoPlayer instance, which isn't ideal.

**Comment by rotvel** [comment]
That makes sense. I've moved the logic back and fixed it in-place.

**Comment by rotvel** [comment]
This breaks playback if a song is added to an empty queue. I'm looking for a fix.

**Comment by dkanada** [comment]
It's probably the small change you made to where playback gets started, the current system is kind of a house of cards. I'm actually looking into it as well to preload the queue when the app is opened since that broke at some point. Something really weird is going on with the callback when a track is started and the next track preload. The queue initialization on app load isn't helping things either.

**Comment by dkanada** [comment]
I think the fix might be proper queue management because I know both ExoPlayer and Chromecast would benefit from that, but I'd want to see if Jellyfin supports a queue over the session player.

**Comment by rotvel** [comment]
I did some testing on master and it turns out I didn't break anything after all. But clearing the queue is broken. After you clear it a "Couldn't play this song" toast appears and if you queue a new song playback doesn't begin. You need clear the queue then restart Gelli before it works again.

So my changes are good to go. The change to playback start works because  exoPlayer.setPlayWhenReady(true) actually does what it says it does, ie it waits until appendDataSource is finished loading the song.

**Comment by rotvel** [comment]
Fixed the issue with clearing the queue. 

There's still a minor cosmetic issue because the position of the song playing (before clearing the queue) is saved and the progress bar of the new song uses this position. But only until playback actually begins, that's why it's a minor issue.

**Comment by rotvel** [comment]
Hmm, clearing the queue doesn't stop playback. It just keeps playing, without any player ui. It also starts to stutter somewhat. The only way to quit playing is by closing Gelli. That can't be right.

**Comment by rotvel** [comment]
The progress issue is also fixed now. I think that's it.

**Comment by dkanada** [comment]
I think this pull request might have reintroduced a race condition with those playback changes. It's pretty rare but I can click on songs and the second item in the queue will start instead of the first. Maybe splitting those changes into a separate pull request so the rest can get merged?

File: app/src/main/java/com/dkanada/gramophone/service/MusicService.java [hunk 1]
```diff
@@ -796,6 +798,7 @@ private void playSongAtImpl(int position) {
     public void pause() {
         pausedByTransientLossOfFocus = false;
         if (playback.isPlaying()) {
+            unRegisterBecomingNoisyReceiver();
```
Comment (dkanada) [inline app/src/main/java/com/dkanada/gramophone/service/MusicService.java:801]: Only comment is that unregister isn't usually `unRegister` in method names.

**Review by dkanada — COMMENTED** [review]
If you could drop the two commits of mine that have snuck in for some reason that would be great. There also seems to be an unnecessary merge commit that could be dropped with no side effects.

**Comment by dkanada** [comment]
This issue is allegedly resolved and was closed recently, so I suppose the pull request can also be closed.

**Pull request closed without merging**